### PR TITLE
Improve junit xml for jenkins

### PR DESCRIPTION
**Pull request by tsondergaard**
This pull request improves the junit xml output in the following ways

* The output isn't significantly different for TEST_CASEs with and without SECTIONs
* With "-#" test results will be grouped by [#filename] in jenkins test results UI
* The test suite name will be used for grouping too

Addressing these three issues ensures that the test results are logically organized in the jenkins test results UI.

This adresses #922 and maybe #320.

I've split the PR up in three tiny commits to ease review.

### CODE REVIEW

**Comment by horenmar** [comment]
I will have to think about this.

In general I agree with the changes, but it is not a change that can be done trivially. Both xml and junit reporters are meant to be consumable by outside tools, so changing their output is potentially breaking change.

File: include/reporters/catch_reporter_junit.hpp [hunk 1]
```diff
@@ -128,6 +128,18 @@ namespace Catch {
             xml.scopedElement( "system-err" ).writeText( trim( stdErrForSuite.str() ), false );
         }
 
+        static std::string fileNameTag( const std::set<std::string> &tags ) {
+            for( std::set<std::string>::const_iterator
+                 it = tags.begin(),
+                 itEnd = tags.end();
+                 it != itEnd;
+                 ++it ) {
+                if( !it->empty() && (*it)[0] == '#' )
```
Comment (horenmar) [inline include/reporters/catch_reporter_junit.hpp]: The performance difference is unlikely to matter, but I would prefer this to be done via `std::set::lower_bound`.

**Comment by tsondergaard** [comment]
This could indeed break things. The best option is perhaps to make a "junit2" reporter with the new behaviour. That is what I have done in my own project. It could also be called "junit-jenkins". The change that encodes the test suite name in the classname attribute as a java package justify this name, I think.

I can alter the pull request to do this.

File: include/reporters/catch_reporter_junit.hpp (hunk 1 above)
Comment (tsondergaard) [inline include/reporters/catch_reporter_junit.hpp]: Okay, done.

File: include/reporters/catch_reporter_junit.hpp [hunk 2]
```diff
@@ -128,6 +128,13 @@ namespace Catch {
             xml.scopedElement( "system-err" ).writeText( trim( stdErrForSuite.str() ), false );
         }
 
+        static std::string fileNameTag( const std::set<std::string> &tags ) {
+            std::set<std::string>::const_iterator it = tags.lower_bound("#");
+            if( it != tags.end() && !it->empty() && it->front() == '#' )
```
Comment (horenmar) [inline include/reporters/catch_reporter_junit.hpp]: `std::string::front()` is C++11 and forward

**Comment by horenmar** [comment]
I thought about it a bit more, and decided to do two things:

1) I am going to merge the changes to the default junit reporter in Catch 2 branch. Since it is a major release, breaking changes are expected and this should be an improvements for most people.

2) I am going to link this PR from reporter documentation, so other people using Jenkins can make the changes for themselves, until Catch 2 is released.

I definitely do not want to add another reporter that is included in Catch by default, and single, relatively small, change is probably not worth a stand-alone reporter if it is going to be merged in (hopefully close) future.

**Comment by tsondergaard** [comment]
I see the changes have been merged to branch dev-modernize, so closing pull request.

**Pull request closed without merging**